exp/operator/controllers/component_customizer.go

```diff
@@ -0,0 +1,318 @@
+		for fg, val := range mSpec.FeatureGates {
+			fgValue = append(fgValue, fg+"="+bool2Str[val])
+		}
+		sort.Strings(fgValue)
+		c.Args = setArg(c.Args, "--feature-gates", strings.Join(fgValue, ","))
+	}
+
+	return c
+}
+
+func setArg(args []string, name, value string) []string {
+	for i, a := range args {
+		if strings.HasPrefix(a, name+"=") {
+			args[i] = name + "=" + value
+			return args
+		}
+	}
+
+	return append(args, name+"="+value)
+}
+
+func removeEnv(envs []corev1.EnvVar, name string) []corev1.EnvVar {
+	for i, a := range envs {
+		if a.Name == name {
+			copy(envs[i:], envs[i+1:])
+			return envs[:len(envs)-1]
+		}
+	}
+
+	return envs
+}
+
+func customizeDeployment(pSpec operatorv1.ProviderSpec, d *appsv1.Deployment) {
+	if pSpec.Deployment != nil {
+		dSpec := pSpec.Deployment
+		if dSpec.Replicas != nil {
+			d.Spec.Replicas = pointer.Int32Ptr(int32(*dSpec.Replicas))
+		}
+		if dSpec.Affinity != nil {
+			d.Spec.Template.Spec.Affinity = dSpec.Affinity
+		}
+		if dSpec.NodeSelector != nil {
+			d.Spec.Template.Spec.NodeSelector = dSpec.NodeSelector
+		}
+		if dSpec.Tolerations != nil {
+			d.Spec.Template.Spec.Tolerations = dSpec.Tolerations
+		}
+
+		for _, pc := range dSpec.Containers {
+			customizeContainer(pc, d)
+		}
+	}
+	// run the customizeManager last so it overrides anything in the deploymentSpec.
+	if pSpec.Manager != nil {
+		for ic, c := range d.Spec.Template.Spec.Containers {
+			if c.Name == managerContainerName {
+				d.Spec.Template.Spec.Containers[ic] = *customizeManager(pSpec.Manager, &d.Spec.Template.Spec.Containers[ic])
+			}
+		}
+	}
+}
+
+// setSpecHashAnnotation computes the hash of the provided spec and sets an annotation of the
+// hash on the provided ObjectMeta.
+func setSpecHashAnnotation(objMeta *metav1.ObjectMeta, spec interface{}) error {
+	jsonBytes, err := json.Marshal(spec)
+	if err != nil {
+		return err
+	}
+	specHash := fmt.Sprintf("%x", sha256.Sum256(jsonBytes))
+	if objMeta.Annotations == nil {
+		objMeta.Annotations = map[string]string{}
+	}
+	objMeta.Annotations[specHashAnnotation] = specHash
+	return nil
+}
+
+func customizeObjectsFn(provider genericprovider.GenericProvider) func(objs []unstructured.Unstructured) ([]unstructured.Unstructured, error) {
+	return func(objs []unstructured.Unstructured) ([]unstructured.Unstructured, error) {
+		results := []unstructured.Unstructured{}
+		for i := range objs {
+			o := objs[i]
+
+			if o.GetKind() == namespaceKind {
+				// filter out namespaces as the targetNamespace already exists as the provider object is in it.
+				continue
+			}
+
+			// Set the owner references so that when the provider is deleted, the components are also deleted.
+			o.SetOwnerReferences(util.EnsureOwnerRef(provider.GetOwnerReferences(), metav1.OwnerReference{
+				APIVersion: operatorv1.GroupVersion.String(),
+				Kind:       provider.GetObjectKind().GroupVersionKind().Kind,
+				Name:       provider.GetName(),
+				UID:        provider.GetUID(),
+			}))
```

Comment:
So initially we were going to rely on this for the clean up. 
But as you know we can't totally because of the clusterrole[bindings]
So we don't need this for cleanup, but isn't it a good purely from an informational PoV?
A user can see straight away that this <resource> is owned and managed by the provider.

Reply:
I have added a finaliser and that does work, so the only followup is "do we remove the ownership?"

Reply:
As far as I can tell this is applying the owner refs to both cluster and namespace scoped objects right? I think, given that we know that applying ORs to cluster scoped objects can be dangerous, we should prevent this from adding ORs to the cluster scoped resources and have the initial implementation lacking in cluster scoped clean up. Else we could cause serious issues if for example, the CRDs get deleted because of an invalid OR